File: ChatGPT/Direct_Attempt/clojure/output/output_10b.clj
;; ChatGPT-generated solution will go here.
(ns output_10b
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(defn parse-grid [lines]
  (mapv #(mapv (fn [ch] (Character/digit ch 10)) (seq %)) lines))

(defn neighbors [[r c] rows cols]
  (filter (fn [[nr nc]] (and (<= 0 nr) (< nr rows) (<= 0 nc) (< nc cols)))
          [[(dec r) c] [(inc r) c] [r (dec c)] [r (inc c)]]))

(defn count-paths [grid start]
  (let [rows (count grid)
        cols (count (first grid))
        memo (atom {})]
    (letfn [(dfs [[r c :as pos] h]
              (if (= h 9)
                1
                (if-let [cached (get @memo [pos h])]
                  cached
                  (let [nxts (for [nbr (neighbors pos rows cols)
                                   :let [nv (get-in grid nbr)]
                                   :when (= nv (inc h))]
                               nbr)
                        total (reduce + (map #(dfs % (inc h)) nxts))]
                    (swap! memo assoc [pos h] total)
                    total))))]
      (dfs start 0))))

(defn part [lines]
  (let [grid (parse-grid lines)
        rows (count grid)
        cols (count (first grid))
        trailheads (for [r (range rows)
                         c (range cols)
                         :when (= 0 (get-in grid [r c]))]
                     [r c])]
    (->> trailheads
         (map #(count-paths grid %))
         (reduce +))))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))

File: ChatGPT/trans_test/clojure_to_python/output/output_10b.py
def parse_grid(lines):
    return [[int(ch) for ch in line] for line in lines]

def neighbors(r, c, rows, cols):
    for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nr, nc = r + dr, c + dc
        if 0 <= nr < rows and 0 <= nc < cols:
            yield nr, nc

def count_paths(grid, start):
    rows, cols = len(grid), len(grid[0])
    memo = {}

    def dfs(r, c, h):
        if h == 9:
            return 1
        if (r, c, h) in memo:
            return memo[(r, c, h)]
        total = 0
        for nr, nc in neighbors(r, c, rows, cols):
            if grid[nr][nc] == h + 1:
                total += dfs(nr, nc, h + 1)
        memo[(r, c, h)] = total
        return total

    r, c = start
    return dfs(r, c, 0)

def part(lines):
    grid = parse_grid(lines)
    rows, cols = len(grid), len(grid[0])
    trailheads = [(r, c) for r in range(rows) for c in range(cols) if grid[r][c] == 0]
    return sum(count_paths(grid, start) for start in trailheads)

def main():
    with open('input.txt') as f:
        lines = f.read().splitlines()
    print(part(lines))

if __name__ == '__main__':
    main()
